### PR TITLE
Build tools for automated trackable deployments.

### PR DESCRIPTION
**Pull request by gmarkall**
This PR requires https://github.com/OpenGamma/OG-Lapack/pull/9 from OG-Lapack.

This consists of tools to combine artifacts from the OG-Maths individual platform builds and its sub projects (i.e. OG-Lapack) and combine them to form a single finished artifact (that we will later dispatch to further testing). 

The verinfo generation tool creates a file that records information about the artifact, including its version, git revision, project name, build numbers for each platform, and similar information from the sub projects. This gets wrapped up in a "blob" (a zip file) with all the artifacts so that it can be moved about easily (e.g. from build slaves to the build master and vice-versa). 

Blobs from individual platform builds are then downloaded by the deploy builder and combined to form a platform-independent blob, that consists of the sources, tests and javadoc jars, as well as a platform independent main jar and a verinfo file that contains information from all the component builds.

Some minor points:
- Coverage is unaffected as these changes only affect the build system
- A bug that was stripping extensions from files included in Jars is fixed.
- The only sensible test of these changes is a successful build from the deploy builder, which I will link to when it has completed (this takes some time due to building absolutely everything on every platform).
- Once this is merged, the buildbot config will need to change so that the testing and master queues pass LAPACK_BUILDDIR instead of LAPACK_EXPORTS to CMake.

Verinfo:

The verinfo file is intended to be easily human- and computer-readable. An example of the verinfo file for a finished build is: 

```
artifacts: [og-maths-0.1.0-SNAPSHOT.jar, og-maths-0.1.0-SNAPSHOT-javadoc.jar, og-maths-0.1.0-SNAPSHOT-sources.jar,
  og-maths-0.1.0-SNAPSHOT-tests.jar]
buildnumber: 30
buildnumbers:
  lnx: 56
  osx: 31
  subprojects:
    OG-Lapack: {lnx: 10, osx: 10, win: 11}
  win: 31
project: OG-Maths
revision: da69cfd5a227bd8c7d01d71e249b6a4e17123ed8
subprojects: {OG-Lapack: 78bb7b167aea5f8b52015610d3ad69f360a84493}
version: 0.1.0-SNAPSHOT
```

This is intended to allow tracing of the build logs, artifacts, and exact revisions used to build any jar that we may have "in the wild" - please comment if you think it is deficient in allowing any of that.

### CODE REVIEW

**Comment by gmarkall** [comment]
Here is the passed "deploy" queue build: http://devmaths-lx-1:8012/builders/deploy/builds/31

**Comment by gmarkall** [comment]
Also the verinfo: http://devmaths-lx-1:8012/blobs/OG-Maths-0.1.0-SNAPSHOT-31.zip-verinfo.yaml
and multiplatform blob of all jars: http://devmaths-lx-1:8012/blobs/OG-Maths-0.1.0-SNAPSHOT-31.zip

**Comment by stuartarchibald** [comment]
Following offline discussion, it was decided that the addition of a time stamp to the `verinfo.yaml` file was a good idea as a backup and a form of build information to which it is easy for humans to relate. Will review following the addition of a patch to perform the described.

**Comment by gmarkall** [comment]
I've added unix timestamps and a human-readable time field to the version info. New deploy queue pass: http://devmaths-lx-1:8012/builders/deploy/builds/32

File: cmake/verinfo.py [hunk 1]
```diff
@@ -4,11 +4,67 @@
 # Please see distribution for license.
 #
 
-import sys, subprocess
+"""This tool creates a version info file for an OG-Maths build on
+the current platform. Version info is a yaml file containing
+the revision number, build number, version number, project name,
+and list of artifacts."""
```
Comment (stuartarchibald) [inline cmake/verinfo.py:10]: Ideally this comment should also now contain `timestamp`, however, in view of this being a textual change I'll add a ticket to patch on next work on this file to save running the entire build again!

File: combine/combine.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        verinfo.close()
+        zipf.close()
+    return verinfos
+
+def validate_version_info(verinfos):
+    """Validate version info - the project, revision, and version should match across
+    blobs and across each subproject in the blobs."""
+
+    def important(attr):
+        """Returns true if it is important for an attribute to be equal across blobs."""
```
Comment (stuartarchibald) [inline combine/combine.py:65]: Nice check.

**Comment by stuartarchibald** [comment]
NOTE: These notes are mainly to hold as process reference. 
- From the best human brain parse I can do, this code seems to do the job for which it was intended.
- Python is certainly a good choice of language for this task.
- Further to my human parse I downloaded the `blob` generated and manually checked it via unzipping and then extracting each jar. The directory structure was then inspected to ensure it contained what is expected. All appeared well.

**Comment by stuartarchibald** [comment]
I am happy to merge this code, it is a large step in the direction of the production jar.
- [[MAT-389]](http://jira.opengamma.com/browse/MAT-389) is created to patch the build system w.r.t. `LAPACK_EXPORTS` needing a change.
- [[MAT-390]](http://jira.opengamma.com/browse/MAT-390) is created to make the minor text change as noted here: https://github.com/OpenGamma/OG-Maths/pull/95#discussion_r12143524

**Pull request merged**